cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Collection, Optional, Sequence, Union, Tuple, List, Type, cast
+
+import copy
+import itertools
+
+import cirq  # pylint: disable=unused-import
+
+
+def flatten(sequence):
+    def _flatten_aux(sequence):
+        if isinstance(sequence, int):
+            yield sequence
+        else:
+            for item in sequence:
+                yield from _flatten_aux(item)
+
+    return tuple(_flatten_aux(sequence))
+
+
+def _from_int(val: int) -> Tuple[Tuple[int, ...], ...]:
+    return ((val,),)
+
+
+def _from_sequence_int(vals: Sequence[int]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple((val,) for val in vals)
+
+
+def _from_sequence_sequence(vals: Sequence[Sequence[int]]) -> Tuple[Tuple[int, ...], ...]:
+    return tuple(tuple(product) for product in vals)
+
+
+class ControlValues:
+    def __init__(
+        self, control_values: Sequence[Union[int, Collection[int], Type['ControlValues']]]
+    ):
+        if len(control_values) == 0:
+            self.vals = cast(Tuple[Tuple[int, ...], ...], (()))
+            self.num_variables = 0
+            self.nxt = None
+            self.itr = None
+            return
+        self.itr = None
+        self.nxt = None
+
+        if len(control_values) > 1:
+            self.nxt = ControlValues(control_values[1:])
+
+        if isinstance(control_values[0], ControlValues):
+            aux = control_values[0].copy()
+            aux.And(self.nxt)
+            self.vals, self.num_variables, self.nxt = aux.vals, aux.num_variables, aux.nxt
+            self.vals = cast(Tuple[Tuple[int, ...], ...], self.vals)
+            return
+
+        val = control_values[0]
+        if isinstance(val, int):
+            self.vals = _from_int(val)
+        elif isinstance(val, (list, tuple)):
+            if isinstance(val[0], int):
+                self.vals = _from_sequence_int(val)
+            else:
+                self.vals = _from_sequence_sequence(val)
+        self.num_variables = len(self.vals[0])
+
+    def And(self, other):  # pylint: disable=invalid-name
```

Comment:
I assume this is camel case to avoid conflict with `and`?  Suggest using `product`.  That would match the pythonic way to do a cartesian product, `itertools.product`